### ISSUE TITLE
Units: remove dimension and scale factor from the definition of Quantity

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#14318]
Currently `Quantity` is constructed as `Quantity(name, dimension, scale_factor, abbrev)`.

The problem: **dimensions** are **not** an **absolute** association to a quantity. For example:

- in natural units length and time are the same dimension (so you can sum meters and seconds).
- SI and cgs units have different dimensions for the same quantities.

At this point a problem arises for scale factor as well: while it is always true that `kilometer / meter == 1000`, some other quantities may have a relative scale factor or not depending on which unit system is currently being used.

You need to handle this somewhere else than in the definition of `Quantity`. I would suggest to move the definitions of relative scale factors and dimensions for `Quantity` objects in the `UnitSystem` objects.

**Comment by schymans** [comment]
I still think that we need the possibility to associate a `Quantity` with a dimension, otherwise I don't see the use of `Quantity`. We could handle dimensional contraction in dimension_system, if we have something like `Quantity(name, dimension, dimension_system, scale_factor, abbrev)`.

**Comment by schymans** [comment]
The deprecation message of `Quantity` states:

> Quantity arguments has been deprecated since SymPy 1.3. Use
SI_quantity_dimension_map instead. See
https://github.com/sympy/sympy/issues/14319 for more info.

Could you provide more info here? E.g. how to use `SI_quantity_dimension_map`?

**Comment by Upabjojr** [comment]
> Could you provide more info here? E.g. how to use SI_quantity_dimension_map?

That is simply a dictionary mapping the quantity to its dimension.

The old API is still working. If it's not, please report it. We need to go through a deprecation process whereby the old code still works, but warns users to change to the new API.

**Comment by Upabjojr** [comment]
Maybe it was an error to change the argument order in `Quantity.__new__`.

**Comment by schymans** [comment]
Yes, it took me a while to adapt my code to the new argument order. It makes more sense, though. 

I still disagree that quantities and dimensions need to be separated. A quantity has no meaning until we define its dimensions, which, of course are specific to the dimension system we use. Could we not define base dimensions and then map those to dimension system-specific dimensions?
I agree that the scale_factor is a different matter, but again, if we define a quantity as a universal constant, than its scale_factor is part of the definition, subject to conversion for different dimension systems.

**Comment by Upabjojr** [comment]
> Could we not define base dimensions and then map those to dimension system-specific dimensions?

The base dimensions are system specific: just think about Maxwell's equations in _SI_ and _cgs_. All their quantities have different dimensions depending on the chose system.

SymPy needs to support all strange dimension systems that arise in theoretical physics, we cannot just stick with SI.

**Comment by schymans** [comment]
I agree that we cannot just stick with just SI, but can't we create a dimension mapping between dimension systems? Then, one could define dimensions of a quantity in the system of choice and then easily switch dimension systems. 

Thinking about your re-factoring, how about keeping the best of both worlds and allow specifying a dimension and dimension system within `Quantity()` for e.g. physical variables, specifying a scale factor within `Quantity()` for e.g. fundamental physical constants, or leaving all of these out of `Quantity()` for e.g. empirical variables where the dimensions are not known a priori? Whatever is specified within the `Quantity()` call should be immutable, whereas if not specified, these attributes can be set dynamically. 

By the way, if we do not have an unambiguous mapping of dimensions between dimension systems, then I do not see the point of using the same name for quantities in different unit systems. For the quantities defined in definitions.py, how do we know which dimension system they relate to? How would we use `planck_length.set_dimension()' if we wanted to do it for e.g. natural dimensions?

**Comment by Upabjojr** [comment]
> can't we create a dimension mapping between dimension systems? Then, one could define dimensions of a quantity in the system of choice and then easily switch dimension systems.

Yes, my idea is that, but I don't have time right now to do that. The class `UnitSystem` needs to be refactored in order to support a mapping.

> Thinking about your re-factoring, how about keeping the best of both worlds and allow specifying a dimension and dimension system within Quantity() for e.g. physical variables, specifying a scale factor within Quantity() for e.g. fundamental physical constants, 

Well, you could potentially subclass `Quantity` and create a specific class, say for example `SIQuantity`, where you can put dimension and scale factor in the constructor. It's actually pretty straightforward to do, you just override `Quantity.__new__`.

> Whatever is specified within the Quantity() call should be immutable, whereas if not specified, these attributes can be set dynamically.

You can subclass _dict_, override `__setitem__` to make it assignable only once. So you can avoid the possibility to vary the scale factor.

> For the quantities defined in definitions.py, how do we know which dimension system they relate to?

As a principle, all quantities can be used in all dimension systems. That's precisely why I removed the dimension parameter from the arguments of `Quantity`.

The dimension system is something introduced by human psychology because we perceive the reality in a way that makes sense to think about certain quantities as being in a certain dimension.

> How would we use `planck_length.set_dimension()' if we wanted to do it for e.g. natural dimensions?

As soon as I have some more time, I'll look into `UnitSystem` and refactor it. The idea is to first load the quantity definitions, then define the unit systems with the mapping to dimensions and scale factors.

Some scale factors can be probably shared among unit systems.

There will probably be a global variable to point to the default unit system, which will most probably be SI.

For other unit systems, my idea is to use context managers and the `with` statement, e.g.:
```python
print(speed_of_light.scale_factor)
with natural_units:
    print(speed_of_light.scale_factor)
```
First print: `299792458`, second print: `1`.

**Comment by schymans** [comment]
These are all very good ideas. Just a quick question: How important is it to switch back and forth between dimension systems? I would think that we usually do most of our work in one dimension system, so the `with` feature might not be used all that much.

**Comment by Upabjojr** [comment]
> How important is it to switch back and forth between dimension systems?

It's not too unreasonable that someone would like to switch from natural units to SI or from cgs to SI. Suppose an example case of someone working with SI but reading a paper using cgs.

**Comment by schymans** [comment]
Fair enough. In this case you would allow setting the unit system globally, e.g. for a whole jupyter notebook, but then switching using the proposed context manager to see the relation between unit systems on a case-by-case basis?

**Comment by Upabjojr** [comment]
> but then switching using the proposed context manager to see the relation between unit systems on a case-by-case basis?

Well, yes. Though it should be advisable to call the `.dimension` and `.scale_factor` methods from an equivalent API to be defined in `UnitSystem`.

**Comment by westurner** [comment]
```python
from sympy import symbols
import pint
ureg = pint.UnitRegistry()

_d1 = 9
_d2 = 18
_m = 31  # Shelves.weight_lb

d1, d2, f1, f2, m, g = symbols('d1 d2 f1 f2 m g')
F1 = (m*g*d1) / (d2 - d1)
print(F1)
values = {
    'd1': _d1 * u.inches,
    'd2': _d2 * u.inches,
    'g': 9.81 * (u.meters / u.seconds**2),  # gravitational constant
    'm': _m * u.pounds
}
print(values)
output = F1.subs(values)
print(output)
```

I think this should work. AFAIU, this used to work; but does not now.

This does not link to an explanation of what needs to be changed:

```
/home/wturner/-wrk/-ce36/carttable/lib/python3.6/site-packages/sympy/physics/units/quantities.py:45: SymPyDeprecationWarning: 

Quantity arguments has been deprecated since SymPy 1.3. Use
SI_quantity_dimension_map instead. See
https://github.com/sympy/sympy/issues/14319 for more info.

  useinstead="SI_quantity_dimension_map",
```

- Why do I need to do manual map lookups? (I've read the issue thread)
- What code do I now need to add in order to use ``SI_quantity_dimension_map``?
- #14319 does not have more info. It has trackback links to e.g. this issue.

**Comment by Upabjojr** [comment]
@westurner I simply realized that it was an error to add the scale factor and dimension to `Quantity`, because it's not an intrinsic property of the quantity. Indeed, different unit systems may associate different dimensions to the same quantity (e.g. electric current in _cgs_ and _SI_).

I haven't found time yet to add the support of different unit systems, but I have added the warning about this behaviour in order to discourage people from using the old API, which was a bad design.

**Comment by schymans** [comment]
I think @westurner actually exposed a problem with `.subs()` and quantities:
```python
from sympy import symbols
import sympy.physics.units as u
m, g = symbols('m g')
values = {
    g: 9.81 * (u.meters / u.seconds**2), 
    m: 3 * u.kilogram
}
print 9.81 * (u.meters / u.seconds**2) * (3 * u.kilogram)
try: 
    (g*m).subs(values)
except Exception as e1:
    print e1
```
returns
```
29.43*kilogram*meter/second**2
expected dimension or 1
```
I do not see the reason why `(g*m).subs(values)` raises an exception, while `9.81 * (u.meters / u.seconds**2) * (3 * u.kilogram)` does not.

**Comment by schymans** [comment]
@westurner I added an example for the new syntax at https://github.com/sympy/sympy/issues/14319

**Comment by Upabjojr** [comment]
> I do not see the reason why (g*m).subs(values) raises an exception, while 9.81 * (u.meters / u.seconds**2) * (3 * u.kilogram) does not.

Blind guess: not a simultaneous substitution.

**Comment by schymans** [comment]
Good point! I never thought about what happens under the hood if a dictionary is passed to `subs()`. In which order are the substitutions done, considering that a dictionary is un-ordered? You are actually right, this is the problem:
```python
(9.81*u.meter*m/u.second**2).subs({m: 3 * u.kilogram})
```
returns `expected dimension or 1`. No idea why.

**Comment by schymans** [comment]
OK, this is probably because `m = meter` in sympy.physics.unit. If I rename `m` to `m1`, there is no problem. However, this seems like a bug to me.

**Comment by asmeurer** [comment]
subs has a flag simultaneous=True that should do a simultaneous substitution. I don't know if it works with units. Also if you use Python 3.6+ dictionaries do keep their order (and otherwise you can use a list of tuples to force the order).

**Comment by schymans** [comment]
Wow, no error with simulataneous=True. I narrowed down the problem:
```python
from sympy import symbols
import sympy.physics.units as u
m = symbols('m')
print 9.81*u.meter
try:
    print (9.81*u.meter).subs({m: 3 * u.kilogram})
except Exception as e1:
    print e1
```
returns
```
9.81*meter
expected dimension or 1
```
while `u.meter == m` returns `False`, so I have no idea why `.subs({m: 3 * u.kilogram})` does anything at all.
